### ISSUE TITLE
13.6.2: compile time error due to ref-safe-context of variable_reference in local_variable_initializer

### ISSUE THREAD

**Issue by KalleOlaviNiemitalo** [dotnet/csharpstandard#910]
**Describe the bug**

In section 13.6.2 of the C# 7.x draft, the following paragraph specifies a "compile time error" that seems impossible to trigger.

> A *local_variable_initializer* for a variable declared `ref` or `ref readonly` shall be of the form “`ref` *variable_reference*”. It is a compile time error if the scope of the local variable is wider than the ref-safe-context of the *variable_reference* (§9.7.2).

**Example**

The paragraph applies to initializing a `ref` local variable, not to ref assignment:

```csharp
ref T local_variable = ref variable_initializer;
```

What would you put as *variable_initializer* to trigger the compile time error?

**Expected behavior**

Remove the sentence about "compile time error", or add an example of invalid code that would trigger it.

**Additional context**

Split off from <https://github.com/dotnet/csharpstandard/issues/899#issuecomment-1684461026>.

**Comment by KalleOlaviNiemitalo** [comment]
The scope of a local variable:

* 13.3.1 (Blocks / General):

  > A block may contain declaration statements (§13.6). The scope of a local variable or constant declared in a block is the block.

* 13.6.2 (Local variable declarations):

  > The scope of a local variable declared in a *local_variable_declaration* is the block in which the declaration occurs. It is an error to refer to a local variable in a textual position that precedes the *local_variable_declarator* of the local variable. Within the scope of a local variable, it is a compile-time error to declare another local variable or constant with the same name.

Can the ref-safe-context of the initializing expression be smaller than the block in which the local variable is declared, without this violating some other constraint?

The ref-safe-context of *variable_reference*:

* 9.7.2.2 (Local variable ref safe context):

  > For a local variable `v`:
  >
  > * If `v` is a reference variable, its ref-safe-context is the same as the ref-safe-context of its initializing expression.
  > * Otherwise its ref-safe-context is the context in which it was declared.

**Comment by KalleOlaviNiemitalo** [comment]
Ugh… from 9.7.2.1 (Ref safe contexts / General):

> * ***declaration-block***: The ref-safe-context of a *variable_reference* to a local variable (§9.2.9) extends from its declaration to the end of the scope in which it is declared, including into any nested *embedded-statements* in this range.

So in this code

```csharp
class C {
    void M() {
        int dummy;
        int target = 0;
        ref int r = ref target;
    }
}
```

the scope of `target` is the whole block, but the ref-safe-context of a *variable_reference* to `target` starts from the declaration of `target`, i.e. does not include the declaration of `dummy`.  The scope of `r` likewise is the whole block, and it is wider than the ref-safe-context of the *variable_reference* `target` in the initializer.  The code is thus invalid according to this paragraph in 13.6.2:

> A *local_variable_initializer* for a variable declared `ref` or `ref readonly` shall be of the form “`ref` *variable_reference*”. It is a compile time error if the scope of the local variable is wider than the ref-safe-context of the *variable_reference* (§9.7.2).

But of course it should not be invalid.  The second sentence of that paragraph is outright wrong.  Changing it to say "the <del>scope</del><ins>ref-safe-context</ins> of the local variable is wider" would remove the error but would make the sentence fully redundant given how 9.7.2.2 defines the ref-safe-context of a local reference variable.

**Comment by jskeet** [comment]
I suspect @gafter will be able to either confirm this or give a cunning counter-example.

**Comment by jskeet** [comment]
Meeting 2023-09-06: Assigning to Nigel who has a potential fix, described in email.

**Comment by jskeet** [comment]
Closing as fixed in #922.
